tracer/src/Datadog.Trace/Ci/Agent/MessagePack/CIEventMessagePackFormatter.cs

```diff
@@ -0,0 +1,199 @@
+        // .
+        private readonly byte[] _eventsBytes = StringEncoding.UTF8.GetBytes("events");
+
+        private readonly ArraySegment<byte> _envelopBytes;
+
+        public CIEventMessagePackFormatter(TracerSettings tracerSettings)
+        {
+            var containerId = ContainerMetadata.GetContainerId();
+            if (containerId is not null)
+            {
+                _containerIdValueBytes = StringEncoding.UTF8.GetBytes(containerId);
+            }
+            else
+            {
+                _containerIdValueBytes = null;
+            }
+
+            var environment = tracerSettings.Environment;
+            if (environment is not null)
+            {
+                _environmentValueBytes = StringEncoding.UTF8.GetBytes(environment);
+            }
+            else
+            {
+                _environmentValueBytes = null;
+            }
+
+            var service = tracerSettings.ServiceName;
+            if (service is not null)
+            {
+                _serviceValueBytes = StringEncoding.UTF8.GetBytes(service);
+            }
+            else
+            {
+                _serviceValueBytes = null;
+            }
+
+            var serviceVersion = tracerSettings.ServiceVersion;
+            if (serviceVersion is not null)
+            {
+                _appVersionValueBytes = StringEncoding.UTF8.GetBytes(serviceVersion);
+            }
+            else
+            {
+                _appVersionValueBytes = null;
+            }
+
+            _envelopBytes = GetEnvelopeArraySegment();
+        }
+
+        public override int Serialize(ref byte[] bytes, int offset, EventsPayload value, IFormatterResolver formatterResolver)
+        {
+            if (value is null)
+            {
+                return 0;
+            }
+
+            var originalOffset = offset;
+
+            // Write envelope
+            MessagePackBinary.EnsureCapacity(ref bytes, offset, _envelopBytes.Count);
+            Buffer.BlockCopy(_envelopBytes.Array, _envelopBytes.Offset, bytes, offset, _envelopBytes.Count);
+            offset += _envelopBytes.Count;
+
+            // Write events
+            if (value.Events.Lock())
+            {
+                var data = value.Events.Data;
+                MessagePackBinary.EnsureCapacity(ref bytes, offset, data.Count);
+                Buffer.BlockCopy(data.Array, data.Offset, bytes, offset, data.Count);
+                offset += data.Count;
+            }
+            else
+            {
+                Log.Error<int>("Error while locking the events buffer with {count} events.", value.Events.Count);
+                offset += MessagePackBinary.WriteNil(ref bytes, offset);
+            }
+
+            return offset - originalOffset;
+        }
+
+        private ArraySegment<byte> GetEnvelopeArraySegment()
+        {
+            var offset = 0;
+            var bytes = new byte[1024];
```

Comment:
Is there a scenario in which we could exceed 1024 bytes? Could an unusually long env / service / hostname cause issues?

Reply:
It can happen, but the `MessagePackBinary` handles the array grow.

As an example, the envelope average in `Datadog.Trace.DuckTyping.Tests` is about 246 bytes.

Reply:
Sounds good, thanks for the clarification